### PR TITLE
[docs] Re-implement website build process

### PR DESCRIPTION
**Pull request by jugglinmike**
This is the basis for the work originally announced in gh-15824. As noted in that issue, we iterated on this patch from [a fork of WPT](https://github.com/bocoup/wpt-docs) in order to avoid interfering with the operation of live documentation site. Most of the changes have been reviewed in that repository; the one exception is the modification of directory names (done to preserve the URL structure of the existing site--see the final commit on this branch).

This implementation is intended to replicate the content of the current site. It includes some improvements (e.g. a search feature and the inclusion of previously-orphaned content), but further improvements can be deferred until after this patch has been approved, merged, and deployed.

After merging, deploying the new site will involve the following changes:

- deactivating the "GitHub Pages" feature for this repository
- configuring https://readthedocs.io to build a site from this repository
- updating the web-platform-tests.org DNS record

According to [our administrative docs](https://web-platform-tests.org/admin/index.html), @jgraham is the only person who can perform that last change, so they should be the one to merge.

### CODE REVIEW

**Comment by Hexcles** [comment]
I'd like to propose a slightly different set of steps to make the transition smoother: 

0. Before merging, create a new `gh-pages` branch. Change GitHub Pages configurations to point at the new branch.
1. Configure https://readthedocs.io to build a site from this repository
2. Update the web-platform-tests.org DNS record (*we might want a small TTL, like 10min*)
3. Deactivate the "GitHub Pages" feature for this repository after a few days (to make sure all DNS caches expire and of course make sure everything is working)

The only risky step would be 2. If anything goes wrong in the new setup, we can revert the DNS back and hopefully it'd take effects within 10min.

**Comment by jugglinmike** [comment]
Very wise, @Hexcles

**Comment by jgraham** [comment]
Wait, what? I didn't realise we were planning to use readthedocs as the backend? Is there some discussion of that?

**Comment by jugglinmike** [comment]
[I first mentioned it when verifying integration on the fork](https://github.com/bocoup/wpt-docs/pull/7#issuecomment-477683356), but I didn't discuss it. Let's do that now.

One of the nice properties of GitHub Pages and Jekyll is that it operates with a "pull" model. That is: once configured, it automatically synchronizes the website with this repository's `master` branch.

Without that, we'd need a dedicated CI task configured with a secret that allowed it to perform the privileged action of publishing the website. We can't do that in TaskCluster, but we can do it in TravisCI.

In order to use a static site generator other than Jekyll (e.g. Sphinx), I think we have three options:

- manually synchronize the website with the `master` branch
- introduce a TravisCI task configured with a secret that grants it push rights to this repository
- use another "pull"-based service (e.g. readthedocs)

I assumed the first alternative is prohibitively inconvenient, and I thought we no longer wanted to rely on TravisCI. That leaves the third option, but maybe there are others. For instance, can this be accomplished with a GitHub Action?

**Comment by jgraham** [comment]
Yeah, this is like the archtypal usecase of a GitHub action.

Do we have an example of the readthedocs hosted site? My initial reaction is that depending on a another site adds an extra layer of risk (it can have downtime that's uncorrelated with GitHub downtime, we probably get an extra set of credentials to manage), but of course there are also advantages compared to rolling our own build step. Does readthedocs impose any additional constraints compared to GH (e.g. in terms of the layout, or requiring advertising on the docs pages)?

**Comment by jugglinmike** [comment]
> Do we have an example of the readthedocs hosted site?

Here's the version currently being built for the documentation fork:

https://wpt-docs.readthedocs.io/en/latest/

> Does readthedocs impose any additional constraints compared to GH (e.g. in terms of the layout, or requiring advertising on the docs pages)?

Not that I'm aware of. It does insert ads by default, but those can be disabled. The ads are ["served ethically"](https://docs.readthedocs.io/en/latest/advertising/ethical-advertising.html) and clearly labeled as such, so I don't take their removal for granted. It'd be nice (though not strictly necessary) to pay for the service if we do remove the ads.

> Yeah, this is like the archtypal usecase of a GitHub action.

Cool. Between readthedocs and GitHub Pages+GitHub Actions, I have a slight preference for readthedocs. Mostly I want to get the website published so I can build on it. Are you comfortable making the call?

**Comment by jgraham** [comment]
> > Do we have an example of the readthedocs hosted site?
> 
> Here's the version currently being built for the documentation fork:
> 
> https://wpt-docs.readthedocs.io/en/latest/
> 
> > Does readthedocs impose any additional constraints compared to GH (e.g. in terms of the layout, or requiring advertising on the docs pages)?
> 
> Not that I'm aware of. It does insert ads by default, but those can be disabled. The ads are ["served ethically"](https://docs.readthedocs.io/en/latest/advertising/ethical-advertising.html) and clearly labeled as such, so I don't take their removal for granted. It'd be nice (though not strictly necessary) to pay for the service if we do remove the ads.

I am fairly strongly against running ads on our documentation pages. I take your point that we should try to pay readthedocs if we do this (and certainly it's a good service that is deserving of our cash), but I think that's going to be logisitically difficult given that we've never managed to directly pay for anything so far (expenditure has either come via company internal budgets or handled by individuals, and getting companies to approve "make an optional donation to this project" is tricky).

> > Yeah, this is like the archtypal usecase of a GitHub action.
> 
> Cool. Between readthedocs and GitHub Pages+GitHub Actions, I have a slight preference for readthedocs. Mostly I want to get the website published so I can build on it. Are you comfortable making the call?

I have a slight preference for GitHub actions. There are several reasons for this

* Fewer different systems. There is already a lot of infrastructure that's hard for even admins/core team members to access, and adding to that list seems undesirable
* Simpler transition. If we can make the action work we can just merge the PR and things should continue to work just as before.
* More flexibility. Building our own docs means that if we want to change or add different tools in the future we aren't limited by what the host supports. 

I understand the concern that this adds a blocker to deploying the changes, but could we at least experiment to find out how hard it is to get this building? There seem to be several docker images available for running sphinx, so it should be possible to build on top of those. Then it's just a matter of scripting the "and now push the built docs to the gh-pages branch" part.

**Comment by jugglinmike** [comment]
I can give it a shot. This branch now has a commit with initial guesswork for how this might work. It's a blind shot, though, and (as I expect you're aware) the GitHub Actions docs aren't so great right now. Does it look right to you? Would you mind if I used [web-platform-tests/wpt-actions-test](https://github.com/web-platform-tests/wpt-actions-test) to experiment? Could you give me admin access to that repository?

**Comment by jugglinmike** [comment]
Thanks @jgraham. I used that demo repository to iterate on the GitHub Action integration, and it seems to be working now.

As you requested in IRC, the script exits early if no relevant files have been modified.

This uses a generic Python Docker image. None of the Sphinx-specific images I found were appropriate for WPT (e.g. based on Python 3, based on some GNU/Linux distribution other than Debian, or too old to support package installation). This means we have to install git and Sphinx with each build, but I think that's acceptable because the overall duration is under 90 seconds.

One disappointing aspect is that commits pushed using the `GITHUB_TOKEN` secret do *not* trigger GitHub Pages to rebuild the site. Instead, it reports the following:

> Your site is having problems building: Page build failed.

Although [the relevant documentation](https://help.github.com/en/articles/troubleshooting-github-pages-builds) only mentions the e-mail address, neither `wpt-pr-bot@noreply.users.github.com` nor my own address satisfied the requirement. I was forced to re-use the token we created for the wpt-pr-bot account. The repository will have to be configured with that value for this patch to function.

File: tools/ci/website_build.sh
Comment (jgraham) [inline tools/ci/website_build.sh]: Note that in general I've been actively trying to move this kind of script from shell into Python, since shell scripts seem to be harder to review an maintain than Python even if they are easier to write in the first place.

File: tools/ci/website_build.sh
Comment (jgraham) [inline tools/ci/website_build.sh]: I think we should roll the "check if the branch is master" part into this, to considerably cut the overhead (i.e. just one action rather than two).

File: tools/ci/website_build.sh
Comment (jgraham) [inline tools/ci/website_build.sh]: It seems nicer to do a shallow clone of the `gh_pages` branch and add the new content there so we get real history of the deployed content rather than overwriting it every time. Is there a reason not to do that?

File: tools/ci/website_build.sh
Comment (jgraham) [inline tools/ci/website_build.sh]: We have now reached the kind of shell magic that leaves me guessing at what's going on :)

**Review by jgraham — COMMENTED** [review]
Looked at the GH actions changes. In general it looks good; I'm happy with the approach, but there are a few comments.

File: tools/ci/website_build.sh
Comment (jugglinmike) [reply tools/ci/website_build.sh]: Sure thing

File: tools/ci/website_build.sh
Comment (jugglinmike) [reply tools/ci/website_build.sh]: Only to avoid the additional complexity in this script. Having a meaningful commit history easily outweighs a few extra statements, though. Plus, this gives us the opportunity to communicate the condition where a patch that satisfies our criteria doesn't alter the built assets (e.g. changes to files in `tools/taskcluster/`).

File: tools/ci/website_build.sh
Comment (jugglinmike) [reply tools/ci/website_build.sh]: This syntax was foreign to me, as well. A multi-line string should do just fine.

**Comment by jugglinmike** [comment]
@jgraham gh-16360 introduced conflicts for this branch, so it will no longer merge cleanly. I've rebased locally; say the word, and I'll force-push the new version.

**Comment by jgraham** [comment]
Please go ahead.

**Comment by jugglinmike** [comment]
All set, @jgraham. [Original branch here.](https://github.com/web-platform-tests/wpt/compare/master...bocoup:docs-sphinx-3-orig)

File: tools/ci/website_build.sh
Comment (jgraham) [inline tools/ci/website_build.sh]: I mean, that's surely not the only reason this can fail, but sure it's good enough for now.

**Review by jgraham — APPROVED** [review]
I really think this script should be in Python, but let's merge this.

File: tools/ci/website_build.sh
Comment (jugglinmike) [reply tools/ci/website_build.sh]: I need to rebase again following gh-16632. Want me to switch to using `set -x` while I'm at it?

**Comment by jugglinmike** [comment]
@jgraham I've rebased again and also added a "fixup" commit to add the "-e" option. Can you take another look?

**Review by jgraham — APPROVED** [review]

**Review by gsnedders — APPROVED** [review]
That fixup LGTM, if you want to clean up the history before we land this.

**Comment by Hexcles** [comment]
@jugglinmike do you have an end-to-end demo of this (Sphinx pages built by GitHub Actions)?

**Comment by jugglinmike** [comment]
> That fixup LGTM, if you want to clean up the history before we land this.

Thanks! Rebased and fixed up the fixup.

> @jugglinmike do you have an end-to-end demo of this (Sphinx pages built by GitHub Actions)?

Yup, [the web-platform-tests/wpt-actions-test repository](https://github.com/web-platform-tests/wpt-actions-test) is configured to use this system to publish to its corresponding GH Pages URL: https://web-platform-tests.github.io/wpt-actions-test/

**Comment by Hexcles** [comment]
@jugglinmike wow, that looks amazing! Great job!

**Comment by gsnedders** [comment]
@jugglinmike https://github.com/web-platform-tests/wpt/runs/126575235 shows this ran on master… but:

> No files related to the website have been modified. Exiting without building.

So we still don't have this. And meanwhile https://web-platform-tests.org is broken.  :(

**Pull request merged**
